scvid/module/incremental_pca.py

```diff
@@ -0,0 +1,121 @@
+    2. *Incremental Learning for Robust Visual Tracking*,
+       D. Ross, J. Lim, R.-S. Lin, M.-H. Yang
+       (https://www.cs.toronto.edu/~dross/ivt/RossLimLinYang_ijcv.pdf)
+
+    Args:
+        g_genes: Number of genes.
+        k_components: Number of principal components.
+        p_oversamples: Additional number of random vectors to sample the range of ``x_ng``
+            so as to ensure proper conditioning.
+        transform: If not ``None`` is used to transform the input data.
+        mean_correct: If ``True`` then the mean correction is applied to the update step.
+            If ``False`` then the data is assumed to be centered and the mean correction
+            is not applied to the update step.
+    """
+
+    def __init__(
+        self,
+        g_genes: int,
+        k_components: int,
+        p_oversamples: int = 10,
+        mean_correct: bool = False,
+        transform: nn.Module | None = None,
+    ) -> None:
+        super().__init__()
+        self.g_genes = g_genes
+        self.k_components = k_components
+        self.p_oversamples = p_oversamples
+        self.mean_correct = mean_correct
+        self.transform = transform
+        self.V_kg: torch.Tensor
+        self.S_k: torch.Tensor
+        self.x_mean_g: torch.Tensor
+        self.x_size: torch.Tensor
+        self.register_buffer("V_kg", torch.zeros(k_components, g_genes))
+        self.register_buffer("S_k", torch.zeros(k_components))
+        self.register_buffer("x_mean_g", torch.zeros(g_genes))
+        self.register_buffer("x_size", torch.tensor(0))
+        self._dummy_param = torch.nn.Parameter(torch.tensor(0.0))
+
+    @staticmethod
+    def _get_fn_args_from_batch(
+        tensor_dict: dict[str, torch.Tensor]
+    ) -> tuple[tuple, dict]:
+        x = tensor_dict["X"]
+        return (x,), {}
+
+    def forward(self, x_ng: torch.Tensor) -> None:
+        if self.transform is not None:
+            x_ng = self.transform(x_ng)
+
+        g = self.g_genes
+        k = self.k_components
+        p = self.p_oversamples
+        m = self.x_size
+        n = x_ng.size(0)
+        assert p + k <= min(n, g)
+
+        # compute SVD of new data
+        if self.mean_correct:
+            x_mean_g = x_ng.mean(dim=0)
+            _, S_q, V_gq = torch.svd_lowrank(x_ng - x_mean_g, q=k + p)
+        else:
+            _, S_q, V_gq = torch.svd_lowrank(x_ng, q=k + p)
+
+        # if not the first batch, merge results
+        if m > 0:
+            SV_kg = torch.diag(S_q[:k]) @ V_gq.T[:k]
+            C = torch.cat([torch.diag(self.S_k) @ self.V_kg, SV_kg], dim=0)
+            if self.mean_correct:
+                mean_correction = (
+                    math.sqrt(m * n / (m + n)) * (self.x_mean_g - x_mean_g)[None, :]
+                )
+                C = torch.cat([C, mean_correction], dim=0)
+            # perform SVD on merged results
+            _, S_q, V_gq = torch.svd_lowrank(C, q=k + p)
+
+        # update buffers
+        self.V_kg = V_gq.T[:k]
```

Comment:
Hmmm... shouldn't with actually be keeping `self.V_qg` and truncate at the very end? I feel like there is something wrong about eager truncation.  With eager truncation, we've effectively working in a k-dim space except for when we're doing the CVD on C, whereas we could be working in a q-dim space by keeping track of oversampled PCs all the way through (but only keeping the first k at the very end).